File: apps/go/design-pattern/command/command_test.go
package command

import (
	"os"
	"path/filepath"
	"testing"
)

func tempPath(t *testing.T, name string) string {
	t.Helper()
	return filepath.Join(t.TempDir(), name)
}

func TestCreateFileCommand(t *testing.T) {
	path := tempPath(t, "test.txt")
	cmd := &CreateFileCommand{Path: path, Content: "hello"}

	if err := cmd.Execute(); err != nil {
		t.Fatalf("Execute 失敗: %v", err)
	}

	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("ファイル読み取り失敗: %v", err)
	}
	if string(data) != "hello" {
		t.Errorf("期待値 'hello', 実際 %q", string(data))
	}
}

func TestCreateFileCommandUndo(t *testing.T) {
	path := tempPath(t, "test.txt")
	cmd := &CreateFileCommand{Path: path, Content: "hello"}

	_ = cmd.Execute()
	if err := cmd.Undo(); err != nil {
		t.Fatalf("Undo 失敗: %v", err)
	}

	if _, err := os.Stat(path); !os.IsNotExist(err) {
		t.Error("Undo 後にファイルが残っている")
	}
}

func TestDeleteFileCommand(t *testing.T) {
	path := tempPath(t, "test.txt")
	_ = os.WriteFile(path, []byte("content"), 0644)

	cmd := &DeleteFileCommand{Path: path}
	if err := cmd.Execute(); err != nil {
		t.Fatalf("Execute 失敗: %v", err)
	}

	if _, err := os.Stat(path); !os.IsNotExist(err) {
		t.Error("削除後にファイルが残っている")
	}
}

func TestDeleteFileCommandUndo(t *testing.T) {
	path := tempPath(t, "test.txt")
	_ = os.WriteFile(path, []byte("content"), 0644)

	cmd := &DeleteFileCommand{Path: path}
	_ = cmd.Execute()

	if err := cmd.Undo(); err != nil {
		t.Fatalf("Undo 失敗: %v", err)
	}

	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("Undo 後のファイル読み取り失敗: %v", err)
	}
	if string(data) != "content" {
		t.Errorf("期待値 'content', 実際 %q", string(data))
	}
}

func TestCompositeCommand(t *testing.T) {
	dir := t.TempDir()
	path1 := filepath.Join(dir, "file1.txt")
	path2 := filepath.Join(dir, "file2.txt")

	cc := &CompositeCommand{
		Commands: []Command{
			&CreateFileCommand{Path: path1, Content: "a"},
			&CreateFileCommand{Path: path2, Content: "b"},
		},
	}

	if err := cc.Execute(); err != nil {
		t.Fatalf("Execute 失敗: %v", err)
	}

	if _, err := os.Stat(path1); os.IsNotExist(err) {
		t.Error("file1 が作成されていない")
	}
	if _, err := os.Stat(path2); os.IsNotExist(err) {
		t.Error("file2 が作成されていない")
	}
}

func TestCompositeCommandUndo(t *testing.T) {
	dir := t.TempDir()
	path1 := filepath.Join(dir, "file1.txt")
	path2 := filepath.Join(dir, "file2.txt")

	cc := &CompositeCommand{
		Commands: []Command{
			&CreateFileCommand{Path: path1, Content: "a"},
			&CreateFileCommand{Path: path2, Content: "b"},
		},
	}

	_ = cc.Execute()
	if err := cc.Undo(); err != nil {
		t.Fatalf("Undo 失敗: %v", err)
	}

	if _, err := os.Stat(path1); !os.IsNotExist(err) {
		t.Error("Undo 後に file1 が残っている")
	}
	if _, err := os.Stat(path2); !os.IsNotExist(err) {
		t.Error("Undo 後に file2 が残っている")
	}
}

func TestCommandDescription(t *testing.T) {
	cmd := &CreateFileCommand{Path: "/tmp/test.txt", Content: ""}
	if cmd.Description() != "ファイル作成: /tmp/test.txt" {
		t.Errorf("Description が期待通りでない: %s", cmd.Description())
	}
}

File: apps/php/design-pattern/tests/CommandTest.php
<?php

declare(strict_types=1);

namespace DesignPattern\Tests;

use DesignPattern\Command\CompositeCommand;
use DesignPattern\Command\CreateFileCommand;
use DesignPattern\Command\DeleteFileCommand;
use PHPUnit\Framework\TestCase;

class CommandTest extends TestCase
{
    private string $testDir;

    protected function setUp(): void
    {
        $this->testDir = sys_get_temp_dir() . '/design_pattern_command_test_' . uniqid();
        mkdir($this->testDir, 0777, true);
    }

    protected function tearDown(): void
    {
        // Clean up test files
        $files = glob($this->testDir . '/*');
        if ($files) {
            array_map('unlink', $files);
        }
        if (is_dir($this->testDir)) {
            rmdir($this->testDir);
        }
    }

    public function testCreateFileCommand(): void
    {
        $path = $this->testDir . '/test.txt';
        $cmd = new CreateFileCommand($path, 'hello');

        $cmd->execute();

        $this->assertFileExists($path);
        $this->assertSame('hello', file_get_contents($path));
    }

    public function testCreateFileCommandUndo(): void
    {
        $path = $this->testDir . '/test.txt';
        $cmd = new CreateFileCommand($path, 'hello');

        $cmd->execute();
        $cmd->undo();

        $this->assertFileDoesNotExist($path);
    }

    public function testDeleteFileCommand(): void
    {
        $path = $this->testDir . '/test.txt';
        file_put_contents($path, 'content');
        $cmd = new DeleteFileCommand($path);

        $cmd->execute();

        $this->assertFileDoesNotExist($path);
    }

    public function testDeleteFileCommandUndo(): void
    {
        $path = $this->testDir . '/test.txt';
        file_put_contents($path, 'original');
        $cmd = new DeleteFileCommand($path);

        $cmd->execute();
        $cmd->undo();

        $this->assertFileExists($path);
        $this->assertSame('original', file_get_contents($path));
    }

    public function testCompositeCommand(): void
    {
        $path1 = $this->testDir . '/a.txt';
        $path2 = $this->testDir . '/b.txt';

        $composite = new CompositeCommand('バッチ作成');
        $composite->addCommand(new CreateFileCommand($path1, 'aaa'));
        $composite->addCommand(new CreateFileCommand($path2, 'bbb'));

        $composite->execute();

        $this->assertFileExists($path1);
        $this->assertFileExists($path2);
    }

    public function testCompositeCommandUndo(): void
    {
        $path1 = $this->testDir . '/a.txt';
        $path2 = $this->testDir . '/b.txt';

        $composite = new CompositeCommand('バッチ作成');
        $composite->addCommand(new CreateFileCommand($path1, 'aaa'));
        $composite->addCommand(new CreateFileCommand($path2, 'bbb'));

        $composite->execute();
        $composite->undo();

        $this->assertFileDoesNotExist($path1);
        $this->assertFileDoesNotExist($path2);
    }

    public function testCommandDescription(): void
    {
        $cmd = new CreateFileCommand('/tmp/x.txt', '');
        $this->assertStringContainsString('ファイル作成', $cmd->getDescription());
    }
}
